File: scripts/convert.py
#!/usr/bin/env python3
"""
convert.py - Extract all resource definitions from a v1.0 Scadable SDK project.

Scans devices/, controllers/, storage/, outbound/ directories for Python files
containing Device, Controller, FileStorage, SQLiteStorage, MQTTOutbound, S3Outbound
subclasses. Outputs a single JSON object with all extracted configs to stdout.

Usage: python3 convert.py <repo_dir>
"""
import sys
import json
import importlib.util
import os
import inspect


def setup_paths(repo_dir):
    """Add repo root and SDK to sys.path so imports resolve."""
    if repo_dir not in sys.path:
        sys.path.insert(0, repo_dir)

    # If scadable isn't installed, check for bundled SDK next to this script
    try:
        import scadable
    except ImportError:
        sdk_dir = os.path.join(os.path.dirname(os.path.dirname(os.path.abspath(__file__))), "sdk")
        if os.path.isdir(sdk_dir) and sdk_dir not in sys.path:
            sys.path.insert(0, sdk_dir)


def load_module(path, module_name):
    """Load a Python module from a file path."""
    spec = importlib.util.spec_from_file_location(module_name, path)
    mod = importlib.util.module_from_spec(spec)
    sys.modules[module_name] = mod
    spec.loader.exec_module(mod)
    return mod


def find_py_files(directory):
    """Find all .py files in a directory (non-recursive)."""
    if not os.path.isdir(directory):
        return []
    return sorted(
        os.path.join(directory, f)
        for f in os.listdir(directory)
        if f.endswith(".py") and not f.startswith("_")
    )


def poll_to_seconds(poll):
    """Convert a poll/run dict to seconds."""
    if poll is None:
        return 0
    if isinstance(poll, dict):
        interval = poll.get("interval", 0)
        unit = poll.get("unit", "seconds")
        if unit == "minutes":
            return interval * 60
        elif unit == "hours":
            return interval * 3600
        return interval
    return int(poll)


def map_connection(conn):
    """Map SDK connection dict to edge-main ConnectionConfig fields."""
    if conn is None:
        return {}, ""

    conn_type = conn.get("type", "")
    result = {}

    if conn_type == "modbus-tcp":
        result["host"] = conn.get("host", "")
        result["port"] = conn.get("port", 502)
        result["slave_id"] = conn.get("slave", 1)
        result["timeout"] = conn.get("timeout", 5.0)
        result["retries"] = conn.get("retries", 3)

    elif conn_type == "modbus-rtu":
        result["serial_port"] = conn.get("port", "")
        result["baudrate"] = conn.get("baud", 9600)
        result["slave_id"] = conn.get("slave", 1)
        result["parity"] = conn.get("parity", "N")
        result["stopbits"] = conn.get("stopbits", 1)
        result["bytesize"] = conn.get("bytesize", 8)
        result["timeout"] = conn.get("timeout", 5.0)

    elif conn_type == "opcua":
        result["host"] = conn.get("host", "")
        result["port"] = conn.get("port", 4840)
        nodes = conn.get("nodes", [])
        result["node_ids"] = [n[1] for n in nodes] if nodes else []
        result["security_policy"] = conn.get("security", "None")
        result["username"] = conn.get("username", "")
        result["password"] = conn.get("password", "")

    elif conn_type == "serial":
        result["serial_port"] = conn.get("port", "")
        result["baudrate"] = conn.get("baud", 115200)
        result["parity"] = conn.get("parity", "N")
        result["stopbits"] = conn.get("stopbits", 1)
        result["bytesize"] = conn.get("bytesize", 8)
        result["timeout"] = conn.get("timeout", 5.0)

    elif conn_type == "ble":
        result["mac"] = conn.get("mac", "")
        result["service_uuid"] = conn.get("service", "")
        chars = conn.get("characteristics", [])
        result["characteristics"] = [{"name": c[0], "uuid": c[1]} for c in chars] if chars else []
        result["scan_timeout"] = conn.get("scan_timeout", 10.0)

    return result, conn_type


def build_decode(cls, conn_type, conn):
    """Build DecodeConfig from registers, fields, or OPC-UA nodes."""
    mappings = []

    # Modbus registers
    registers = getattr(cls, "registers", None)
    if registers:
        for reg in registers:
            mappings.append({
                "from": f"reg_{reg.address}",
                "to": reg.name,
                "scale": getattr(reg, "scale", 1.0) or 1.0,
                "offset": getattr(reg, "offset", 0.0) or 0.0,
            })

    # Serial fields
    fields = getattr(cls, "fields", None)
    if fields:
        for field in fields:
            mapping = {
                "from": f"field_{field.start}_{field.length}",
                "to": field.name,
                "scale": getattr(field, "scale", 1.0) or 1.0,
                "offset": getattr(field, "offset", 0.0) or 0.0,
                "field_type": getattr(field, "type", ""),
                "start": field.start,
                "length": field.length,
            }
            mappings.append(mapping)

    # OPC-UA node name mappings
    if conn_type == "opcua" and conn:
        nodes = conn.get("nodes", [])
        for name, node_id in (nodes or []):
            mappings.append({
                "from": node_id,
                "to": name,
                "scale": 1.0,
                "offset": 0.0,
            })

    if not mappings:
        return None

    return {
        "transform_type": "declarative",
        "mappings": mappings,
    }


def extract_devices(modules):
    """Extract Device subclasses from loaded modules."""
    from scadable.device import Device

    devices = []
    for mod in modules:
        for name in dir(mod):
            obj = getattr(mod, name)
            if not isinstance(obj, type) or not issubclass(obj, Device) or obj is Device:
                continue

            conn_raw = getattr(obj, "connection", None)
            connection, conn_type = map_connection(conn_raw)
            frequency = poll_to_seconds(getattr(obj, "poll", None))
            decode = build_decode(obj, conn_type, conn_raw)

            device = {
                "device_id": obj.id,
                "protocol": conn_type,
                "frequency": frequency,
                "connection": connection,
            }
            if decode:
                device["decode"] = decode

            # Extract historian config if present
            historian = getattr(obj, "historian", None)
            if historian is not None:
                hist_config = {
                    "fields": historian.fields or [],
                    "condition": getattr(historian, "condition", "all") or "all",
                }
                hist_interval = getattr(historian, "interval", None)
                if hist_interval is not None:
                    hist_config["interval"] = poll_to_seconds(hist_interval)
                else:
                    hist_config["interval"] = frequency  # default to device poll rate
                device["historian"] = hist_config

            devices.append(device)

    return devices


def extract_storage(modules):
    """Extract FileStorage/SQLiteStorage subclasses."""
    from scadable.storage import FileStorage, SQLiteStorage

    results = []
    for mod in modules:
        for name in dir(mod):
            obj = getattr(mod, name)
            if not isinstance(obj, type):
                continue
            if obj in (FileStorage, SQLiteStorage):
                continue

            if issubclass(obj, FileStorage):
                storage_type = "file"
            elif issubclass(obj, SQLiteStorage):
                storage_type = "sqlite"
            else:
                continue

            results.append({
                "id": obj.id,
                "storage_type": storage_type,
                "path": obj.path,
                "max_size": obj.max_size,
                "warning_threshold": getattr(obj, "warn_at", 80),
            })

    return results


def extract_outbound(modules):
    """Extract MQTTOutbound/S3Outbound subclasses."""
    from scadable.outbound import MQTTOutbound, S3Outbound
    from scadable.device import Device

    results = []
    for mod in modules:
        for name in dir(mod):
            obj = getattr(mod, name)
            if not isinstance(obj, type):
                continue
            if obj in (MQTTOutbound, S3Outbound):
                continue

            if issubclass(obj, S3Outbound):
                outbound_type = "s3"
            elif issubclass(obj, MQTTOutbound):
                outbound_type = "mqtt"
            else:
                continue

            # Resolve device class references to IDs
            device_ids = []
            for dev in (obj.devices or []):
                if isinstance(dev, type) and issubclass(dev, Device):
                    device_ids.append(dev.id)
                elif isinstance(dev, str):
                    device_ids.append(dev)

            entry = {
                "id": obj.id,
                "outbound_type": outbound_type,
                "devices": device_ids,
            }

            # S3-specific fields
            if outbound_type == "s3":
                storage_cls = getattr(obj, "storage", None)
                if storage_cls and isinstance(storage_cls, type):
                    entry["storage"] = storage_cls.id
                else:
                    entry["storage"] = ""
                entry["prefix"] = getattr(obj, "prefix", "") or ""
                entry["max_age"] = getattr(obj, "max_age", "") or ""

            results.append(entry)

    return results


def extract_controllers(modules, base_dir):
    """Extract Controller subclasses."""
    from scadable.controller import Controller
    from scadable.device import Device

    results = []
    for mod in modules:
        for name in dir(mod):
            obj = getattr(mod, name)
            if not isinstance(obj, type) or not issubclass(obj, Controller) or obj is Controller:
                continue

            # Resolve uses to device IDs
            use_ids = []
            for dev in (obj.uses or []):
                if isinstance(dev, type) and issubclass(dev, Device):
                    use_ids.append(dev.id)
                elif isinstance(dev, str):
                    use_ids.append(dev)

            interval = poll_to_seconds(getattr(obj, "run", None))

            # Get source file path relative to repo root
            source_file = ""
            mod_file = getattr(mod, "__file__", "")
            if mod_file and base_dir:
                source_file = os.path.relpath(mod_file, base_dir)

            results.append({
                "id": obj.id,
                "interval": interval,
                "uses": use_ids,
                "source_file": source_file,
            })

    return results


def load_directory(repo_dir, dirname):
    """Load all .py modules from a directory, returning the loaded modules."""
    directory = os.path.join(repo_dir, dirname)
    modules = []
    for path in find_py_files(directory):
        basename = os.path.splitext(os.path.basename(path))[0]
        module_name = f"{dirname}.{basename}"
        try:
            mod = load_module(path, module_name)
            modules.append(mod)
        except Exception as e:
            print(f"Error loading {path}: {e}", file=sys.stderr)
            raise
    return modules


def main():
    if len(sys.argv) < 2:
        print("Usage: convert.py <repo_dir>", file=sys.stderr)
        sys.exit(1)

    repo_dir = os.path.abspath(sys.argv[1])
    setup_paths(repo_dir)

    result = {
        "devices": [],
        "storage": [],
        "outbound": [],
        "controllers": [],
    }

    errors = []

    # Load modules from each directory
    # Order matters: devices first (controllers/outbound reference them)
    try:
        device_modules = load_directory(repo_dir, "devices")
        storage_modules = load_directory(repo_dir, "storage")
        outbound_modules = load_directory(repo_dir, "outbound")
        controller_modules = load_directory(repo_dir, "controllers")
    except Exception as e:
        print(json.dumps({"error": str(e)}), file=sys.stderr)
        sys.exit(1)

    # Extract
    try:
        result["devices"] = extract_devices(device_modules)
        result["storage"] = extract_storage(storage_modules)
        result["outbound"] = extract_outbound(outbound_modules)
        result["controllers"] = extract_controllers(controller_modules, repo_dir)
    except Exception as e:
        print(json.dumps({"error": str(e)}), file=sys.stderr)
        sys.exit(1)

    total = sum(len(v) for v in result.values())
    if total == 0:
        print("Warning: no resource definitions found", file=sys.stderr)

    print(json.dumps(result))


if __name__ == "__main__":
    main()

File: internal/converter/convert.go
package converter

import (
	"encoding/json"
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"runtime"
)

// ConvertResult holds all extracted resources from a v1.0 SDK project.
type ConvertResult struct {
	Devices     []DeviceConfig     `json:"devices"`
	Storage     []StorageConfig    `json:"storage"`
	Outbound    []OutboundConfig   `json:"outbound"`
	Controllers []ControllerConfig `json:"controllers"`
}

// DeviceConfig represents a device extracted from Python.
type DeviceConfig struct {
	DeviceID   string           `json:"device_id" yaml:"device_id"`
	Protocol   string           `json:"protocol" yaml:"protocol"`
	Frequency  int              `json:"frequency" yaml:"frequency"`
	Filter     []string         `json:"filter,omitempty" yaml:"filter,omitempty"`
	Connection ConnectionConfig `json:"connection" yaml:"connection"`
	Decode     *DecodeConfig    `json:"decode,omitempty" yaml:"decode,omitempty"`
}

// ConnectionConfig represents device connection parameters.
type ConnectionConfig struct {
	// Modbus TCP
	Host    string  `json:"host,omitempty" yaml:"host,omitempty"`
	Port    int     `json:"port,omitempty" yaml:"port,omitempty"`
	Timeout float64 `json:"timeout,omitempty" yaml:"timeout,omitempty"`
	Retries int     `json:"retries,omitempty" yaml:"retries,omitempty"`

	// Shared (Modbus TCP + RTU)
	SlaveID int `json:"slave_id,omitempty" yaml:"slave_id,omitempty"`

	// Serial / Modbus RTU
	SerialPort string `json:"serial_port,omitempty" yaml:"serial_port,omitempty"`
	Baudrate   int    `json:"baudrate,omitempty" yaml:"baudrate,omitempty"`
	Parity     string `json:"parity,omitempty" yaml:"parity,omitempty"`
	Stopbits   int    `json:"stopbits,omitempty" yaml:"stopbits,omitempty"`
	Bytesize   int    `json:"bytesize,omitempty" yaml:"bytesize,omitempty"`

	// OPC-UA
	NodeIDs        []string `json:"node_ids,omitempty" yaml:"node_ids,omitempty"`
	SecurityPolicy string   `json:"security_policy,omitempty" yaml:"security_policy,omitempty"`
	Username       string   `json:"username,omitempty" yaml:"username,omitempty"`
	Password       string   `json:"password,omitempty" yaml:"password,omitempty"`

	// BLE
	MAC             string              `json:"mac,omitempty" yaml:"mac,omitempty"`
	ServiceUUID     string              `json:"service_uuid,omitempty" yaml:"service_uuid,omitempty"`
	Characteristics []BLECharacteristic `json:"characteristics,omitempty" yaml:"characteristics,omitempty"`
	ScanTimeout     float64             `json:"scan_timeout,omitempty" yaml:"scan_timeout,omitempty"`
}

// BLECharacteristic represents a BLE GATT characteristic to read.
type BLECharacteristic struct {
	Name string `json:"name" yaml:"name"`
	UUID string `json:"uuid" yaml:"uuid"`
}

// DecodeConfig describes how to transform raw device data.
type DecodeConfig struct {
	TransformType string         `json:"transform_type" yaml:"transform_type"`
	Mappings      []FieldMapping `json:"mappings" yaml:"mappings"`
	WasmModule    string         `json:"wasm_module,omitempty" yaml:"wasm_module,omitempty"`
}

// FieldMapping is a single field transform rule.
type FieldMapping struct {
	From      string  `json:"from" yaml:"from"`
	To        string  `json:"to" yaml:"to"`
	Scale     float64 `json:"scale" yaml:"scale"`
	Offset    float64 `json:"offset" yaml:"offset"`
	FieldType string  `json:"field_type,omitempty" yaml:"field_type,omitempty"`
	Start     *int    `json:"start,omitempty" yaml:"start,omitempty"`
	Length    *int    `json:"length,omitempty" yaml:"length,omitempty"`
}

// StorageConfig represents a storage backend extracted from Python.
type StorageConfig struct {
	ID               string `json:"id" yaml:"id"`
	StorageType      string `json:"storage_type" yaml:"storage_type"`
	Path             string `json:"path" yaml:"path"`
	MaxSize          uint64 `json:"max_size" yaml:"max_size"`
	WarningThreshold int    `json:"warning_threshold" yaml:"warning_threshold"`
}

// OutboundConfig represents an outbound destination extracted from Python.
type OutboundConfig struct {
	ID           string   `json:"id" yaml:"id"`
	OutboundType string   `json:"outbound_type" yaml:"outbound_type"`
	Devices      []string `json:"devices" yaml:"devices"`
	Storage      string   `json:"storage,omitempty" yaml:"storage,omitempty"`
	Prefix       string   `json:"prefix,omitempty" yaml:"prefix,omitempty"`
	MaxAge       string   `json:"max_age,omitempty" yaml:"max_age,omitempty"`
}

// ControllerConfig represents a controller extracted from Python.
type ControllerConfig struct {
	ID         string   `json:"id" yaml:"id"`
	Interval   int      `json:"interval" yaml:"interval"`
	Uses       []string `json:"uses" yaml:"uses"`
	SourceFile string   `json:"source_file" yaml:"source_file"`
}

// ConvertPython runs the Python conversion script and returns all resource configs.
func ConvertPython(repoDir string) (*ConvertResult, error) {
	scriptPath := findScript()
	if scriptPath == "" {
		return nil, fmt.Errorf("convert.py not found")
	}

	cmd := exec.Command("python3", scriptPath, repoDir)
	cmd.Stderr = os.Stderr

	out, err := cmd.Output()
	if err != nil {
		return nil, fmt.Errorf("python conversion failed: %w", err)
	}

	var result ConvertResult
	if err := json.Unmarshal(out, &result); err != nil {
		return nil, fmt.Errorf("failed to parse converter output: %w (output: %s)", err, string(out))
	}

	// Validate devices
	for i, d := range result.Devices {
		if d.DeviceID == "" {
			return nil, fmt.Errorf("device %d: device_id is required", i)
		}
		if d.Protocol == "" {
			return nil, fmt.Errorf("device %d (%s): protocol is required", i, d.DeviceID)
		}
		if d.Frequency <= 0 {
			return nil, fmt.Errorf("device %d (%s): frequency must be > 0", i, d.DeviceID)
		}
		switch d.Protocol {
		case "modbus-tcp":
			if d.Connection.Host == "" {
				return nil, fmt.Errorf("device %s: modbus-tcp requires connection.host", d.DeviceID)
			}
		case "modbus-rtu", "serial":
			if d.Connection.SerialPort == "" {
				return nil, fmt.Errorf("device %s: %s requires connection.serial_port", d.DeviceID, d.Protocol)
			}
		case "opcua":
			if d.Connection.Host == "" {
				return nil, fmt.Errorf("device %s: opcua requires connection.host", d.DeviceID)
			}
		case "ble":
			// BLE can scan by MAC or service UUID, both optional
		}
	}

	// Validate storage
	for i, s := range result.Storage {
		if s.ID == "" {
			return nil, fmt.Errorf("storage %d: id is required", i)
		}
		if s.Path == "" {
			return nil, fmt.Errorf("storage %d (%s): path is required", i, s.ID)
		}
		if s.MaxSize == 0 {
			return nil, fmt.Errorf("storage %d (%s): max_size must be > 0", i, s.ID)
		}
	}

	// Validate outbound
	for i, o := range result.Outbound {
		if o.ID == "" {
			return nil, fmt.Errorf("outbound %d: id is required", i)
		}
		if o.OutboundType != "mqtt" && o.OutboundType != "s3" {
			return nil, fmt.Errorf("outbound %d (%s): outbound_type must be 'mqtt' or 's3'", i, o.ID)
		}
	}

	// Validate controllers
	for i, c := range result.Controllers {
		if c.ID == "" {
			return nil, fmt.Errorf("controller %d: id is required", i)
		}
		if c.Interval <= 0 {
			return nil, fmt.Errorf("controller %d (%s): interval must be > 0", i, c.ID)
		}
	}

	return &result, nil
}

// DriverName maps a protocol to its driver binary name.
func DriverName(protocol string) string {
	switch protocol {
	case "modbus-tcp", "modbus-rtu":
		return "driver-modbus"
	case "opcua":
		return "driver-opcua"
	case "serial":
		return "driver-serial"
	case "ble":
		return "driver-ble"
	case "mqtt":
		return "driver-mqtt"
	default:
		return "driver-" + protocol
	}
}

// findScript locates convert.py by checking multiple paths.
func findScript() string {
	candidates := []string{
		"/usr/local/bin/convert.py",
		"scripts/convert.py",
	}

	// Also check relative to this source file (for tests run from subdirectories)
	_, thisFile, _, ok := runtime.Caller(0)
	if ok {
		root := filepath.Dir(filepath.Dir(filepath.Dir(thisFile)))
		candidates = append(candidates, filepath.Join(root, "scripts", "convert.py"))
	}

	for _, p := range candidates {
		if _, err := os.Stat(p); err == nil {
			return p
		}
	}
	return ""
}
